softwareComponents/stm32cxx/src/drivers/gpio.hpp

```diff
@@ -70,6 +70,24 @@ struct Gpio: public Peripheral< GPIO_TypeDef >, public detail::Gpio< Gpio > {
             return *this;
         }
 
+        Pin& setupODAlternate( bool pull, bool pull_up = true ) {
+            port().enableClock();
+
+            LL_GPIO_InitTypeDef cfg = { };
+
+            cfg.Pin = 1 << _pos;
+            cfg.Mode = LL_GPIO_MODE_ALTERNATE;
+            cfg.Speed = LL_GPIO_SPEED_FREQ_LOW;
+            cfg.OutputType = LL_GPIO_OUTPUT_OPENDRAIN;
+            cfg.Pull = !pull ?
+                LL_GPIO_PULL_NO :
+                pull_up ? LL_GPIO_PULL_UP : LL_GPIO_PULL_DOWN;
+            cfg.Alternate = LL_GPIO_AF_6;
+            LL_GPIO_Init( _periph, &cfg );
+
+            return *this;
+        }
+
```

Comment:
Toto je velmi nesystémové a tato funkce by neměla existovat. Piny má 0-16 alternativních funkcí a ty si zde "ad-hoc" zvolíš funkci, která ti zrovna pasuje na I2C. Systémové řešení, které používá celá stm32cxx, je inicializace jako OD pomocí `setupODOutput` a následně až periferie si nastaví správnou alternativní funkci pinu. Protože jen periferie ví, která funkce je třeba.

Reply:
Moje myšlenka byla, ale jak vidím, tak jsem na to zapomněl, že by tato funkce na stavovala různé alternativní funkce, které tomu dá periferie (popřípadně jiné parametry). Chtěl jsem, aby přímé nastavování bylo co nejvíce schováno, aby šlo dle potřeby upravovat na jednom místě, ale nevím, jak je to s přenositelností. Možná jsem to špatně pochopil.

Reply:
Můj pohled na věc: GPIO mají dva stavy/konfigurace:
- to, jak elektricky interagují s vnějším světem (rychlost náběžných hran, pull-up/down, push-pull/open-drain)
- to, co pin "krmi signálem" zevnitř procesoru.

Zatímco to, čím krmím nožičku záleží na použité periferii, tak to, jak elektricky interaguje ven je na periferii nezávislé. Konkrétní případ - I2C můžu mít s interním pull-upem nebo bez. Ale perifierii je to jedno.

Ano, ta hranice je mlhavá; zrovna u I2C nedává smysl push-pull výstup. I přesto mi dává smysl přes GPIO konfigurovat jen vnější interakci a přes periferie zdroj signálu.